sni.go

```diff
@@ -17,25 +17,118 @@ package tcpproxy
 import (
 	"bufio"
 	"bytes"
+	"context"
 	"crypto/tls"
 	"io"
 	"net"
+	"strings"
 )
 
 // AddSNIRoute appends a route to the ipPort listener that says if the
 // incoming TLS SNI server name is sni, the connection is given to
 // dest. If it doesn't match, rule processing continues for any
 // additional routes on ipPort.
 //
+// By default, the proxy will route all ACME tls-sni-01 challenges
+// received on ipPort to all SNI dests. You can disable ACME routing
+// with StopACME.
+//
 // The ipPort is any valid net.Listen TCP address.
 func (p *Proxy) AddSNIRoute(ipPort, sni string, dest Target) {
-	p.addRoute(ipPort, sniMatch(sni), dest)
+	cfg := p.configFor(ipPort)
+	if !cfg.stopACME {
+		if len(cfg.acmeTargets) == 0 {
+			p.addRoute(ipPort, (*acmeMatch)(&cfg.acmeTargets))
+		}
+		cfg.acmeTargets = append(cfg.acmeTargets, dest)
+	}
+
+	p.addRoute(ipPort, sniMatch{sni, dest})
+}
+
+// AddStopACMESearch prevents ACME probing of subsequent SNI routes.
+func (p *Proxy) AddStopACMESearch(ipPort string) {
+	p.configFor(ipPort).stopACME = true
+}
+
+type sniMatch struct {
+	sni    string
+	target Target
+}
+
+func (m sniMatch) match(br *bufio.Reader) Target {
+	if clientHelloServerName(br) == string(m.sni) {
+		return m.target
+	}
+	return nil
+}
+
+type acmeMatch []Target
+
+func (m *acmeMatch) match(br *bufio.Reader) Target {
```

Comment:
pointer to slice is an odd receiver type.

I see why you did it, but it looks weird both here and on line 41.

It might be more clear as:

```go
// acmeMatch matches "*.acme.invalid" ACME challenges and searches
// for a route in cfg.acmeTargets to see which backend has the challenge
// response.
type acmeMatch struct { cfg *Config }
```

And then above:

```go
if len(cfg.acmeTargets) == 0 {
   p.addRoute(ipPort, acmeMatch{cfg})
}
```

And this is:

```go
func (m acmeMatch) match(br *bufio.Reader) Target {
    targets := m.cfg.acmeTargets
...
```